### PR TITLE
KMP Algorithm in Java

### LINKED ISSUE #878: KMP Algorithm in C++, Python, Java
I would like to work on this issue

### PR DESCRIPTION
**Pull request by Bhairavi-shah**
Fixes #878 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Sample Input / Output.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

#### Changes proposed in this pull request:
- Added code for KMP Algorithm in Java

#### Languages Used:
- Java

#### Files Added:
- Knuth_Morris_Pratt_Algorithm/KMP.java

### CODE REVIEW

**Comment by marwahmanbir** [comment]
Please fix and complete the checklist above

**Comment by Bhairavi-shah** [comment]
Have completed the checklist. Please review my code.

**Comment by marwahmanbir** [comment]
Some of the points are still not checked. Remove the bullets from checklist points and check them with [x]

File: Knuth_Morris_Pratt_Algorithm/KMP.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
+/* Given a text txt[0..n-1] and a pattern pat[0..m-1], we need to find all occurrences of pat[] in txt[] */
+/* Assume: n>m */
+
```
Comment (marwahmanbir) [inline Knuth_Morris_Pratt_Algorithm/KMP.java]: Remove this extra line.

File: Knuth_Morris_Pratt_Algorithm/KMP.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        int M = pat.length();
+        int N = txt.length();
+        int lps[] = new int[M]; // longest_prefix_suffix array
+        int j = 0; // index for pat[]
+
+        computeLPSArray(pat, M, lps); //calculate lps[] array
+
+        int i = 0; //index for txt[]
+        while(i<N)
```
Comment (marwahmanbir) [inline Knuth_Morris_Pratt_Algorithm/KMP.java]: Add space around ``<``.

File: Knuth_Morris_Pratt_Algorithm/KMP.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            if(pat.charAt(j) == txt.charAt(i))
```
Comment (marwahmanbir) [inline Knuth_Morris_Pratt_Algorithm/KMP.java]: Space after ``if``.

File: Knuth_Morris_Pratt_Algorithm/KMP.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            {
+                j++;
+                i++;
+            }
+
+            if(j==M)
```
Comment (marwahmanbir) [inline Knuth_Morris_Pratt_Algorithm/KMP.java]: Space around ``==``.

File: Knuth_Morris_Pratt_Algorithm/KMP.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            {
+                System.out.println("Found pattern at index"+(i-j));
+                j = lps[j - 1];
+            }
+            else if(i<N && pat.charAt(j) != txt.charAt(i))
```
Comment (marwahmanbir) [inline Knuth_Morris_Pratt_Algorithm/KMP.java]: Space after ``if`` and around ``<``.

File: Knuth_Morris_Pratt_Algorithm/KMP.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            {
+                if(j != 0)
+                    j = lps[j -1];
```
Comment (marwahmanbir) [inline Knuth_Morris_Pratt_Algorithm/KMP.java]: Fix spacing around ``-`` operator.

Comment (marwahmanbir) [inline Knuth_Morris_Pratt_Algorithm/KMP.java]: Fix spacing around ``-`` operator.

File: Knuth_Morris_Pratt_Algorithm/KMP.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+
+    void computeLPSArray(String pat, int M, int lps[]) // length of the previous longest prefix suffix
+    {
+        int len = 0;
+        int i = 1;
+        lps[0] = 0; //lps[0] is always 0
+
+        while(i<M) //calculate lps[i] for i=1 to M-1
```
Comment (marwahmanbir) [inline Knuth_Morris_Pratt_Algorithm/KMP.java]: Space around ``<``.

File: Knuth_Morris_Pratt_Algorithm/KMP.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            if(pat.charAt(i) == pat.charAt(len))
```
Comment (marwahmanbir) [inline Knuth_Morris_Pratt_Algorithm/KMP.java]: After ``if``.

File: Knuth_Morris_Pratt_Algorithm/KMP.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    }
+}
+
+/* Output
+
+Pattern found at 1
+Pattern found at 7
+Pattern found at 16
+
+*/
```
Comment (marwahmanbir) [inline Knuth_Morris_Pratt_Algorithm/KMP.java]: Remove line 80, add one at the end of file instead.

**Comment by marwahmanbir** [comment]
Good work :+1:

**Pull request merged**